mlflow/utils/proto_json_utils.py

```diff
@@ -370,36 +397,54 @@ def parse_tf_serving_input(inp_dict, schema=None):
-                    raise MlflowException(
-                        "Failed to parse input data. This model contains a tensor-based model"
-                        " signature with input names, which suggests a dictionary input mapping"
-                        f" input name to tensor, but an input of type {type(input_data)} was found."
-                    )
-                type_dict = dict(zip(schema.input_names(), schema.numpy_types()))
-                for col_name in input_data.keys():
-                    input_data[col_name] = np.array(
-                        input_data[col_name], dtype=type_dict.get(col_name)
-                    )
+                types_dict = schema.input_dict()
+                if (
+                    len(types_dict) == 1
+                    and isinstance(input_data, list)
+                    and not any(isinstance(x, dict) for x in input_data)
+                ):
+                    # for schemas with a single column (not List[Dict]), match input with column
+                    input_data = {list(types_dict.keys())[0]: input_data}
+            # Un-named schema should only contain a single column
+            elif not isinstance(input_data, list):
+                raise MlflowException(
+                    "Failed to parse input data. This model contains an un-named tensor-based"
+                    " model signature which expects a single n-dimensional array as input,"
+                    f" however, an input of type {type(input_data)} was found."
+                )
+        if isinstance(input_data, dict):
+            input_data = {
+                col: convert_data_type(data, types_dict.get(col))
+                for col, data in input_data.items()
+            }
+        elif isinstance(input_data, list):
+            # List of dictionaries of column_name -> value mapping
+            # List[Dict] must correspond to a schema with named columns
+            if all(isinstance(x, dict) for x in input_data):
+                input_data = [
+                    {
+                        col: convert_data_type(value, types_dict.get(col))
+                        for col, value in data.items()
+                    }
+                    for data in input_data
+                ]
+            # List of values
             else:
-                if not isinstance(input_data, list):
-                    raise MlflowException(
-                        "Failed to parse input data. This model contains an un-named tensor-based"
-                        " model signature which expects a single n-dimensional array as input,"
-                        f" however, an input of type {type(input_data)} was found."
-                    )
-                input_data = np.array(input_data, dtype=schema.numpy_types()[0])
+                spec = schema.inputs[0] if schema else None
+                input_data = convert_data_type(input_data, spec)
         else:
-            if isinstance(input_data, dict):
```

Comment:
Can we make sure to preserve this optimistic structural casting logic if the schema is not defined and the input structure is a dict?

Reply:
Yep, it's here https://github.com/mlflow/mlflow/pull/10094/files#r1370985629

Reply:
I need to get my prescription checked. I had thought that the added lines were nested at indent 10 (under the schema presence encapsulation conditional), not at the same level (8). 
looks good!